Restore TextureUploader

zakorgy opened the pull request:
Fixes: #148
My first attempt to restore the TextureUploader API.
The differences between the original API logic and this implementation:
- It uses PBO-s to upload textures, we have an upload buffer for each image to do the same. Since OpenGL doesn't have to know the size when creating these buffers, it can create PBO-s separately from textures. In our case we must know the size for the buffers and the image creation is when we receive this data.
-  It splits the buffer upload and the texture update in two separate steps. Since our upload buffers and images are paired together, we do the upload and update in the final step. If we want to follow the original API logic, we should move some input parameters between the buffer upload and texture update function .
- The original API doesn't need to store a reference to the data in the UploadChunk structure, because the data is uploaded to the corresponding PBO when we create an UploadChunk. We only upload it in the final step: `update_impl`, so until then we have store a [reference](https://github.com/szeged/webrender/compare/master...zakorgy:restore-texture-uploader#diff-d058594280164bb23d9b1ca25ba9fe79R3215) to the data in the UploadChunk.

dati91 commented:
To be honest, not sure how to approach this, because, in vulkan, you always do the buffer to image copy, what the pbo part tries to achieve. Also we don't expose our buffers like OpenGL's pbo.
And we always batch our calls, and submit before we want to draw.

So our default behavior is probably what the `UploadMethod::PixelBuffer` does and i'm not sure we can support the `Immediate` part, because we need a buffer to write into a texture afaik.

I think the main part what you should focus on, is that the `renderer.rs` can call the "same" device.rs function (like you do now), and under the hood, OpenGL will do its own thing, and we will do the "vulkan version" of it in our implementation.

zakorgy commented:
I placed an `unimplemented!()` to the `Intermediate` path, and simplified the TextureUploader API:
 - Removed the unused structures
 - Moved the `upload_texture` steps in the `upload` call, so our logic doesn't change.
